test/gurka/test_use_living_streets.cc

```diff
@@ -0,0 +1,94 @@
+#include "gurka.h"
+#include "test.h"
+#include <gtest/gtest.h>
+
+using namespace valhalla;
+
+const std::vector<std::string>& costing = {"auto",          "hov",        "taxi",
+                                           "bus",           "bicycle",    "truck",
+                                           "motor_scooter", "motorcycle", "pedestrian"};
+
+void validate_path(const valhalla::Api& result, const std::vector<std::string>& expected_names) {
+  ASSERT_EQ(result.trip().routes(0).legs_size(), 1);
+  auto leg = result.trip().routes(0).legs(0);
+  gurka::assert::raw::expect_path(result, expected_names);
+}
+
+class LivingStreetTest : public ::testing::Test {
+protected:
+  static gurka::map use_living_streets_map;
+
+  static void SetUpTestSuite() {
+    const std::string ascii_map = R"(
+    A-1----B--------E-----2-F
+           |        |       |
+           |        |       |
+           |        |       3
+           |        |       |
+           C--------D       G
+                     \
+                      I
+    )";
+
+    const gurka::ways ways = {
+        {"AB", {{"highway", "living_street"}}}, {"BE", {{"highway", "living_street"}}},
+        {"EF", {{"highway", "living_street"}}}, {"FG", {{"highway", "residential"}}},
+        {"BC", {{"highway", "residential"}}},   {"CD", {{"highway", "residential"}}},
+        {"DE", {{"highway", "residential"}}},   {"DI", {{"highway", "secondary"}}},
+    };
+
+    const auto layout = gurka::detail::map_to_coordinates(ascii_map, 100);
+    use_living_streets_map =
+        gurka::buildtiles(layout, ways, {}, {}, "test/data/gurka_use_living_streets");
+
+    // set the same historical speeds for all roads
+    test::customize_historical_traffic(use_living_streets_map.config, [](baldr::DirectedEdge& e) {
+      e.set_free_flow_speed(40);
+      e.set_constrained_flow_speed(40);
+      return std::vector<int16_t>{};
+    });
+  }
+};
+
+gurka::map LivingStreetTest::use_living_streets_map = {};
+
+TEST_F(LivingStreetTest, test_default_value) {
+  for (const auto& c : costing)
+    if (c == "bicycle" || c == "pedestrian")
+      // living_street tag shouldn't affect these costings too much
+      validate_path(gurka::route(use_living_streets_map, "1", "2", c), {"AB", "BE", "EF"});
+    else
+      // avoid living_streets by default; use living_streets only if the route starts or ends at
+      // 'living_street' edge
+      validate_path(gurka::route(use_living_streets_map, "1", "2", c),
+                    {"AB", "BC", "CD", "DE", "EF"});
+}
```

Comment:
This test is great, easy to read and follow.

Reply:
All credit goes to @genadz and his tests on track avoidance :)